File: fsharp/sieve/Sieve.fs
module Sieve

    let private cons (head:'a) (tail:'a list):'a list = head::tail

    let rec private sieve (differenceList:int list -> int list) (numbers:int list) =
        match numbers with
        | [] ->
            []
            |> differenceList
        | (head:int)::(tail:int list) ->
            tail
            |> List.filter (fun (element:int) -> element % head <> 0)
            |> sieve (differenceList << (cons head))

    let primes (limit:int):int list = sieve id [2..limit]

File: solutions/scala/sieve/1/src/main/scala/Sieve.scala
object Sieve
{

    private def sieve(stack:List[Int], numbers:List[Int]):List[Int] =
        numbers match
        case Nil => stack.reverse
        case (head:Int)::(tail:List[Int]) => sieve(head::stack, tail.filter((element:Int) => element % head != 0))

    def primes(limit:Int):List[Int] = sieve(Nil, (2 to limit).toList)
}